### ISSUE TITLE
cURL connection issues to some repos when using HTTPS

### ISSUE THREAD

**Issue by nikgoodley-ibboost** [git-for-windows/git#2924]
- [X] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

This bug report is for all versions of Git for Windows greater than 2.28.0, starting from 2.29.0.rc0.windows.1. It is still an issue in 2.29.2.2. 64-bit versions used throughout.

```
$ git --version --build-options

$ git --version --build-options
git version 2.29.0.rc0.windows.1
cpu: x86_64
built from commit: 58a43b46e3811645a1fe75703627576b0a97f7eb
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh

```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

Windows 10 Pro, but replicated on Windows Server 2019 (AWS EC2 instance) and Windows Server 2012 (physical) - all 64-bit

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.19042.630]

```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

I've tried with defaults and customised versions; none of these have an effect.

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"

Editor Option: VIM
Custom Editor Path:
Default Branch Option: main
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Core
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled

```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

It's also been observed now on a Debian docker image using Git 2.29.2.1 inside an Ubuntu AWS EC2 instance, so it may not be strictly related to Windows however the change notes include the update from cURL from 7.71.1 to 7.72 in 2.29.rc0 which seems to coincide with the problem.

All invocations have been either from eu-west2 (London) AWS region, a London office or a home in SE England which doesn't affect frequency of problem.

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Git Bash, but also through Jenkins or CMD.

 - What commands did you run to trigger this issue?

```
# A single run will not always replicate the issue, but the basic form is:
git init
git fetch -- https://github.com/github/site-policy

# About 15% of requests fails so best to run this in a loop in an empty directory
REPO=https://github.com/github/site-policy
ITERS=20; COUNT=0; for IT in $(seq $ITERS); do git init; git fetch -- ${REPO}; if [ $? -ne 0 ]; then COUNT=$((COUNT+1)); fi; done; echo "Total failures: ${COUNT} from ${ITERS} runs"

```
 - What did you expect to occur after running these commands?

Fetching to complete without problems

 - What actually happened instead?

About 15% of executions (almost always between 10-20%) an error like the below occurs. Sometimes it's a slightly different message.
```
error: RPC failed; curl 18 transfer closed with outstanding read data remaining
fatal: expected flush after ref listing
```

When running with Git tracing on whilst on one of these machines I see

```
22:52:22.512480 http.c:774              == Info: Found bundle for host github.com: 0x8afbc0 [serially]
22:52:22.512480 http.c:774              == Info: Can not multiplex, even if we wanted to!
22:52:22.512480 http.c:774              == Info: Re-using existing connection! (#0) with host github.com
22:52:22.512480 http.c:774              == Info: Connected to github.com (140.82.121.3) port 443 (#0)
22:52:22.512480 http.c:774              == Info: Server auth using Basic with user 'REDACTED'
22:52:22.513480 http.c:721              => Send header, 0000000391 bytes (0x00000187)
22:52:22.513480 http.c:733              => Send header: POST /REDACTED/git-upload-pack HTTP/1.1
22:52:22.513480 http.c:733              => Send header: Host: github.com
22:52:22.513480 http.c:733              => Send header: Authorization: Basic <redacted>
22:52:22.513480 http.c:733              => Send header: User-Agent: git/2.29.2.windows.2
22:52:22.513480 http.c:733              => Send header: Accept-Encoding: deflate, gzip, br, zstd
22:52:22.513480 http.c:733              => Send header: Content-Type: application/x-git-upload-pack-request
22:52:22.513480 http.c:733              => Send header: Accept: application/x-git-upload-pack-result
22:52:22.513480 http.c:733              => Send header: Git-Protocol: version=2
22:52:22.513480 http.c:733              => Send header: Content-Length: 128
22:52:22.513480 http.c:733              => Send header:
22:52:22.513480 http.c:774              == Info: upload completely sent off: 128 out of 128 bytes
22:52:22.866159 http.c:774              == Info: Mark bundle as not supporting multiuse
22:52:22.866159 http.c:721              <= Recv header, 0000000017 bytes (0x00000011)
22:52:22.866159 http.c:733              <= Recv header: HTTP/1.1 200 OK
22:52:22.866159 http.c:721              <= Recv header, 0000000026 bytes (0x0000001a)
22:52:22.866159 http.c:733              <= Recv header: Server: GitHub Babel 2.0
22:52:22.866159 http.c:721              <= Recv header, 0000000052 bytes (0x00000034)
22:52:22.866159 http.c:733              <= Recv header: Content-Type: application/x-git-upload-pack-result
22:52:22.866159 http.c:721              <= Recv header, 0000000028 bytes (0x0000001c)
22:52:22.866159 http.c:733              <= Recv header: Transfer-Encoding: chunked
22:52:22.866159 http.c:721              <= Recv header, 0000000040 bytes (0x00000028)
22:52:22.866159 http.c:733              <= Recv header: Expires: Fri, 01 Jan 1980 00:00:00 GMT
22:52:22.866159 http.c:721              <= Recv header, 0000000018 bytes (0x00000012)
22:52:22.866159 http.c:733              <= Recv header: Pragma: no-cache
22:52:22.866159 http.c:721              <= Recv header, 0000000053 bytes (0x00000035)
22:52:22.866159 http.c:733              <= Recv header: Cache-Control: no-cache, max-age=0, must-revalidate
22:52:22.866159 http.c:721              <= Recv header, 0000000023 bytes (0x00000017)
22:52:22.866159 http.c:733              <= Recv header: Vary: Accept-Encoding
22:52:22.866159 http.c:721              <= Recv header, 0000000023 bytes (0x00000017)
22:52:22.866159 http.c:733              <= Recv header: X-Frame-Options: DENY
22:52:22.866159 http.c:721              <= Recv header, 0000000056 bytes (0x00000038)
22:52:22.866159 http.c:733              <= Recv header: X-GitHub-Request-Id: E9CA:122F1:1FE87F:2C596F:5FC96C26
22:52:22.866159 http.c:721              <= Recv header, 0000000002 bytes (0x00000002)
22:52:22.866159 http.c:733              <= Recv header:
22:52:22.872165 http.c:774              == Info: TLSv1.3 (IN), TLS alert, close notify (256):
22:52:22.872165 http.c:774              == Info: transfer closed with outstanding read data remaining
22:52:22.872165 http.c:774              == Info: Closing connection 0
22:52:22.872165 http.c:774              == Info: TLSv1.3 (OUT), TLS alert, close notify (256):
error: RPC failed; curl 18 transfer closed with outstanding read data remaining
fatal: expected flush after ref listing
```

whereas in a working iteration you see

```
...
22:52:37.209136 http.c:774              == Info: upload completely sent off: 106 out of 106 bytes
22:52:37.543028 http.c:774              == Info: Mark bundle as not supporting multiuse
22:52:37.543028 http.c:721              <= Recv header, 0000000017 bytes (0x00000011)
22:52:37.543028 http.c:733              <= Recv header: HTTP/1.1 200 OK
22:52:37.543028 http.c:721              <= Recv header, 0000000026 bytes (0x0000001a)
22:52:37.543028 http.c:733              <= Recv header: Server: GitHub Babel 2.0
22:52:37.543028 http.c:721              <= Recv header, 0000000052 bytes (0x00000034)
22:52:37.543028 http.c:733              <= Recv header: Content-Type: application/x-git-upload-pack-result
22:52:37.543028 http.c:721              <= Recv header, 0000000028 bytes (0x0000001c)
22:52:37.543028 http.c:733              <= Recv header: Transfer-Encoding: chunked
22:52:37.543028 http.c:721              <= Recv header, 0000000040 bytes (0x00000028)
22:52:37.543028 http.c:733              <= Recv header: Expires: Fri, 01 Jan 1980 00:00:00 GMT
22:52:37.543028 http.c:721              <= Recv header, 0000000018 bytes (0x00000012)
22:52:37.543028 http.c:733              <= Recv header: Pragma: no-cache
22:52:37.543028 http.c:721              <= Recv header, 0000000053 bytes (0x00000035)
22:52:37.543028 http.c:733              <= Recv header: Cache-Control: no-cache, max-age=0, must-revalidate
22:52:37.543028 http.c:721              <= Recv header, 0000000023 bytes (0x00000017)
22:52:37.543028 http.c:733              <= Recv header: Vary: Accept-Encoding
22:52:37.543028 http.c:721              <= Recv header, 0000000023 bytes (0x00000017)
22:52:37.543028 http.c:733              <= Recv header: X-Frame-Options: DENY
22:52:37.543028 http.c:721              <= Recv header, 0000000055 bytes (0x00000037)
22:52:37.543028 http.c:733              <= Recv header: X-GitHub-Request-Id: E9D0:BE5D:203961:2CA53B:5FC96C34
22:52:37.543028 http.c:721              <= Recv header, 0000000002 bytes (0x00000002)
22:52:37.543028 http.c:733              <= Recv header:
22:52:37.553039 http.c:774              == Info: Connection #0 to host github.com left intact
22:52:37.556043 http.c:774              == Info: Found bundle for host github.com: 0x222fbc0 [serially]
22:52:37.556043 http.c:774              == Info: Can not multiplex, even if we wanted to!
22:52:37.556043 http.c:774              == Info: Re-using existing connection! (#0) with host github.com
22:52:37.556043 http.c:774              == Info: Connected to github.com (140.82.121.3) port 443 (#0)
22:52:37.556043 http.c:774              == Info: Server auth using Basic with user 'REDACTED'
22:52:37.557037 http.c:721              => Send header, 0000000391 bytes (0x00000187)
22:52:37.557037 http.c:733              => Send header: POST /REDACTED/git-upload-pack HTTP/1.1
22:52:37.557037 http.c:733              => Send header: Host: github.com
22:52:37.557037 http.c:733              => Send header: Authorization: Basic <redacted>
22:52:37.557037 http.c:733              => Send header: User-Agent: git/2.29.2.windows.2
22:52:37.557037 http.c:733              => Send header: Accept-Encoding: deflate, gzip, br, zstd
22:52:37.557037 http.c:733              => Send header: Content-Type: application/x-git-upload-pack-request
22:52:37.557037 http.c:733              => Send header: Accept: application/x-git-upload-pack-result
22:52:37.557037 http.c:733              => Send header: Git-Protocol: version=2
22:52:37.557037 http.c:733              => Send header: Content-Length: 597
22:52:37.557037 http.c:733              => Send header:
22:52:37.557037 http.c:774              == Info: upload completely sent off: 597 out of 597 bytes
22:52:37.930839 http.c:774              == Info: Mark bundle as not supporting multiuse
22:52:37.930839 http.c:721              <= Recv header, 0000000017 bytes (0x00000011)
22:52:37.930839 http.c:733              <= Recv header: HTTP/1.1 200 OK
22:52:37.930839 http.c:721              <= Recv header, 0000000026 bytes (0x0000001a)
22:52:37.930839 http.c:733              <= Recv header: Server: GitHub Babel 2.0
22:52:37.930839 http.c:721              <= Recv header, 0000000052 bytes (0x00000034)
22:52:37.930839 http.c:733              <= Recv header: Content-Type: application/x-git-upload-pack-result
22:52:37.930839 http.c:721              <= Recv header, 0000000028 bytes (0x0000001c)
22:52:37.930839 http.c:733              <= Recv header: Transfer-Encoding: chunked
22:52:37.930839 http.c:721              <= Recv header, 0000000040 bytes (0x00000028)
22:52:37.930839 http.c:733              <= Recv header: Expires: Fri, 01 Jan 1980 00:00:00 GMT
22:52:37.930839 http.c:721              <= Recv header, 0000000018 bytes (0x00000012)
22:52:37.930839 http.c:733              <= Recv header: Pragma: no-cache
22:52:37.930839 http.c:721              <= Recv header, 0000000053 bytes (0x00000035)
22:52:37.930839 http.c:733              <= Recv header: Cache-Control: no-cache, max-age=0, must-revalidate
22:52:37.930839 http.c:721              <= Recv header, 0000000023 bytes (0x00000017)
22:52:37.930839 http.c:733              <= Recv header: Vary: Accept-Encoding
22:52:37.930839 http.c:721              <= Recv header, 0000000023 bytes (0x00000017)
22:52:37.931841 http.c:733              <= Recv header: X-Frame-Options: DENY
22:52:37.931841 http.c:721              <= Recv header, 0000000055 bytes (0x00000037)
22:52:37.931841 http.c:733              <= Recv header: X-GitHub-Request-Id: E9D0:BE5D:2039B1:2CA669:5FC96C35
22:52:37.931841 http.c:721              <= Recv header, 0000000002 bytes (0x00000002)
```

I have traced in Wireshark and found there's not RST packet in the exchanges with the Github servers (two IPs observed from server side, no pattern to the failures according to IP however) but it's not clear to me why this would be.

My suspicion is it's the update to cURL 7.72 bundled into this release but going through the change list it's not apparent how I'd be able to test these individually against what seems to be some specific conditions of the target repo (and server?).

https://curl.se/changes.html#7_72_0

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

This is probably the most curious thing; it happens reliably (in terms of frequency) but on a minority of repositories. Working with private repos I can't share is where I've done most of the analysis, but going through Github public repos, I have found four out of the 17 I checked where the issue happens. In all four of these, when fetching 20 times there were between 3 and 5 failures. In the other 13 there are no occurrences at all in 20 iterations on each.

**Issue seen:**
github/covid19-dashboard
github/site-policy
github/VisualStudio
github/platform-samples

**No issue seen:**
git/git
github/training-kit
github/octodns
github/explore
github/dmca
github/deli
github/textarea-autosize
github/octobox
github/securitylab
github/graphql-client
github/ossar-action
github/scientist
github/Spoon-Knife
github/testrepo

**Comment by dscho** [comment]
> the update from cURL from 7.71.1 to 7.72 in 2.29.rc0 which seems to coincide with the problem.

Could you verify that by replacing the `mingw64\bin\libcurl-4.dll` file in an unpacked v2.29.0-rc0 with a supposedly working version from a previous Git for Windows?

**Comment by nikgoodley-ibboost** [comment]
Good idea. That seems to put that theory to rest then, as it made no change on the outcome (I also tried doing the same with curl.exe for giggles, and running the test in a new session without impact):

```
cp /c/Temp/gh/PortableGit-2.28.0/mingw64/bin/libcurl-4.dll /c/Temp/gh/PortableGit-2.29.0-rc0/mingw64/bin/libcurl-4.dll
...
Total failures: 6 from 20 runs

cp /c/Temp/gh/PortableGit-2.28.0/mingw64/bin/curl.exe /c/Temp/gh/PortableGit-2.29.0-rc0/mingw64/bin/curl.exe 
...
error: RPC failed; curl 18 transfer closed with outstanding read data remaining
fatal: expected flush after ref listing
Total failures: 7 from 20 runs
```

**Comment by nikgoodley-ibboost** [comment]
Going further down that copy-from-good rabbithole, doing a merge between 2.28.0 and 2.29.0.rc0 I found that (for this https repo) it comes back solely to **mingw64\libexec\git-core\git-remote-https.exe**

Replacing in an existing 2.29.2(2) version just this one file from 2.28.0 was sufficient to not see the problem in 50 runs.

Presumably the other git-remote-*tp*.exe may have the same problem since they seem to have identical file sizes.

**Comment by dscho** [comment]
I looked at the diff between v2.28.0.windows.1 and v2.29.0-rc0.windows.1 in http.c and remote-curl.c and saw nothing that could explain that regression. Hmm.

**Comment by nikgoodley-ibboost** [comment]
Hi Johannes, I saw that 2.30 was out (and I had a new VM that automatically got it) so I wanted to try to see whether ths issue is fixed. However, on none of our infrastructure (VM, physical, cloud-based docker containers, of all OSs, in three different locations) can we replicate now with 2.29(.2 or otherwise).

I tried on both our problematic internal repo and three of the ones I mentioned in the ticket with the problem and they were all fine on hundreds of attempts with the same set of users.

So this is very confusing. The only can I think of is that this was somehow an issue specifically on the GitHub side where potentially they were routing _some_ traffic to _some_ repos only for client versions 2.29.x (or specifically, only for the 2.29.x versions of git-remote-https...) to something that wasn't working. I couldn't find much information about what they change at a granular level nor bug reports so I guess it's unlikely to ever know what the background was to it.

Anyway from my perspective I can no longer replicate on machines that reliably replicated previously so I consider this a Chrismas gift and happy to close. Thanks for your consideration and wish you a lovely 2021 (can't be worse?!)
